crypto/fipsmodule/rand/fork_detect.c

```diff
@@ -42,37 +42,102 @@ DEFINE_STATIC_MUTEX(g_fork_detect_lock)
 DEFINE_BSS_GET(volatile char *, g_fork_detect_addr)
 DEFINE_BSS_GET(uint64_t, g_fork_generation)
 DEFINE_BSS_GET(int, g_ignore_madv_wipeonfork)
+DEFINE_BSS_GET(int, g_ignore_pthread_atfork)
 
+static int init_fork_detect_madv_wipeonfork(void *addr, long page_size) {
+
+  // Some versions of qemu (up to at least 5.0.0-rc4, see linux-user/syscall.c)
+  // ignore |madvise| calls and just return zero (i.e. success). But we need to
+  // know whether MADV_WIPEONFORK actually took effect. Therefore try an invalid
+  // call to check that the implementation of |madvise| is actually rejecting
+  // unknown |advice| values.
+  if (madvise(addr, (size_t)page_size, -1) == 0 ||
+      madvise(addr, (size_t)page_size, MADV_WIPEONFORK) != 0) {
+    munmap(addr, (size_t)page_size);
+    return 0;
```

Comment:
I think having the returned value 1 for error and 0 for success may be less confusing to me, especially with the other functions such as pthread_atfork() and madvise() doing that.
Is there a defined SUCCESS return value?

Reply:
AWS-LC use the convention of 1 meaning success throughout. Starting to change that locally would be confusing imo.

We don't have a "success" macro. While such a macro would be neat, it would also tradeoff easier upstream merges etc.